### ISSUE TITLE
ReadyToRun images are not as efficient for .NET Core 3 as NGen for .NET Framework

### ISSUE THREAD

**Issue by AlexChuev** [dotnet/wpf#1748]
* .NET Core Version: 3.0 Preview 8
* Windows version: 1809
* Does the bug reproduce also in WPF for .NET Framework 4.8?: No
* Is this bug related specifically to tooling in Visual Studio (e.g. XAML Designer, Code editing, etc...)? No

 **Problem description:**  
Since Ngen.exe and Native Image Task are not available in .NET Core 3, ReadyToRun images are the only way to reduce the application startup time and avoid delays caused by JIT compilation. However, our tests show that for the same WPF application that initializes three times faster after using Ngen.exe, ReadyToRun images provide the difference of only 25%.  
This is a huge hit for our users, since with many libraries and theme resources, JIT compilation is one of the main factors affecting load times. For some real-life applications, the use of Ngen.exe allowed our users to shave up to 6 seconds (half) off the initial start time.
 
 **Minimal repro:**  
 You can find samples for .NET Core 3 and .NET Framework here: https://github.com/AlexChuev/ReadyToRunPerformanceTest  
 These samples use DevExpress WPF assemblies to demonstrate how Ngen.exe and ReadyToRun images affect projects with many classes and theme resources. If you need published apps or other samples for your tests, please let me know.

P.S. I'm posting this in the WPF repo because the difference between Ngen.exe and ReadyToRun images highly affects WPF applications. Normally, the JIT compiler processes classes and methods only when they are needed for the program execution. However, when WPF loads theme resources for a control, this causes all classes referenced in these resources (including classes referenced in currently unused or invisible parts) to be processed by the JIT compiler. In addition, static constructors of many WPF classes contain the DependencyProperty registration code that may cause even more classes and methods to be JITted.

**Comment by rladuca** [comment]
@fadimounir Where should we file a companion bug on .NET?

**Comment by fadimounir** [comment]
@rladuca Sure you can file a companion bug on dotnet/coreclr. We'd be interested in diagnosing this more, although one thing to keep in mind is that R2R by design principle will always be a bit slower than fragile native images, because they are version resilient (they do not have the same fragility as old ngen images).

> If you need published apps or other samples for your tests, please let me know.

@AlexChuev That would be very helpful

cc @jkotas FYI

**Comment by grubioe** [comment]
New CoreCLR issues created, keeping open in WPF repo for tracking purposes.

**Comment by lindexi** [comment]
The other startup performance:

![](https://user-images.githubusercontent.com/16054566/168516872-a02129f3-26a4-45c8-9e1b-343beea0c72a.png)

**Comment by ekalchev** [comment]
Please look at this issue. It basically negates all performance benefits from .net core for desktop apps making it slower than .net framework. We are currently migrating .net framework WPF app to .net core and we were unpleasantly surprised that it got slower startup times. This issue affect the entire class of desktop application.

**Comment by lindexi** [comment]
Should we modified the milestones from 7.0 to 9.0?

**Comment by pchaurasia14** [comment]
@lindexi - This one would require some in-depth investigations at our end and I don't think we'll be able to address this for .NET 9. I'm marking this to future release for now and will change it to .NET 9 if we can get to this sooner.

**Comment by ekalchev** [comment]
> @lindexi - This one would require some in-depth investigations at our end and I don't think we'll be able to address this for .NET 9. I'm marking this to future release for now and will change it to .NET 9 if we can get to this sooner.

Honestly, I don't know what are you planning for .net9 but it will be irrelevant if ppl stop using WPF. And this is one of the most important properties of a desktop app - startup time. When .net core progress, we expected startup time to improve because of all the optimizations promised by .net runtime, but it got worse and nobody do anything about it - for years!

I hoped you'll fix it in .net6, now the best estimate is .net10? Lets face it - we just need to move away from that shit WPF and stop waiting for any improvements. I don't know what is so important for .net9 that you are planning that takes precedence of embarrassingly slow XAML parser. May be some Win11 styling that we can do by yourself but not the fix for fu cking XAML parser that we can't fix my our own.

**Comment by simonegli8** [comment]
There should also be an option to create Ready2Run assemblies in place after deployment as a replacement for ngen.exe. It's a huge disadvantage that Ready2Run can only be used at build time and not after an application has been deployed as it was with ngen.exe. The main disadvantage of JIT is the slow startup times, and there should be a way to avoid that without compromising the advatage of JIT that it can run on any platform & CPU. With Read2Run, one has to compile for each platform & CPU. This should be done after deployment, not when publishing

**Comment by ekalchev** [comment]
> There should also be an option to create Ready2Run assemblies in place after deployment as a replacement for ngen.exe. It's a huge disadvantage that Ready2Run can only be used at build time and not after an application has been deployed as it was with ngen.exe. The main disadvantage of JIT is the slow startup times, and there should be a way to avoid that without compromising the advatage of JIT that it can run on any platform & CPU. With Read2Run, one has to compile for each platform & CPU. This should be done after deployment, not when publishing

I strongly disagree that Ready2Run should function as ngen. I've used both for years and R2R is better technology (apart from that it is not as efficient as ngen, but that has nothing to do with compile time or deployment time native compilation)

When ngen is run on the client machine, you don't have access to native code debug symbols that can be generated only on the client machine. For desktop application that is huge, since you can't debug memory dumps from client machine crash. Also our experience with ngen shows that it slowed installation time by 300% on some machines. These two factors are huge drawback of ngen.

Most desktop application include some parts of native code components and that force you to provide packages for all supported platforms anyway and the in the reality it is very rare to be able to provide single package that can run on all platforms. That why you still see platform dependent installers on many .net apps. My personal opinion is that the whole idea of JIT and how it promised to emit faster code that native binaries didn't age well. You can see languages like Swift and Go abandon the idea and went with native compilation. I hope .net follow this and NativeAOT to become the future of .net.
